src/spec.c

```diff
@@ -2644,6 +2697,8 @@ int IndexSpec_UpdateDoc(IndexSpec *spec, RedisModuleCtx *ctx, RedisModuleString
     return REDISMODULE_ERR;
   }
 
+  RedisSearchCtx_LockSpecWrite(&sctx);
```

Comment:
Can be called from BG thread from `Indexes_ScanProc` or from main thread on key notification,
both have the GIL, so it is needed to lock for write?

Reply:
yes, because each modification of the index can affect the readers that might be running in the background. We can discuss the necessity of locking for reading from the main thread or the BG indexer, but I think as a first step we should lock in every case and then we'll benchmark the effect and decide if it is negligible or not.

Reply:
I suggest to document assumptions/guidelines
Such as the possibility to skip locking for read when holding the GIL and only reading data which GC doesn't touch, such as field names.

Reply:
When are benchmarks planned for?

Reply:
In a couple of weeks probably